src/mlpack/methods/xgboost/xgboost_tree_regressor.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file methods/xgboost/xgboost_tree_regressor.hpp
+ * @author Rishabh Garg
+ *
+ * Definition of the xgboost tree regressor class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_XGBOOST_XGBOOST_TREE_REGRESSOR_HPP
+#define MLPACK_METHODS_XGBOOST_XGBOOST_TREE_REGRESSOR_HPP
+
+#include <mlpack/methods/decision_tree/decision_tree_regressor.hpp>
+#include <mlpack/methods/decision_tree/multiple_random_dimension_select.hpp>
+
+namespace mlpack {
+namespace ensemble {
+
+/**
+ * The XGboostTreeRegressor class provides the implementation of Gradient
+ * Boosted Decision Tree Regressor as described in the XGBoost paper:
+ *
+ * @code
+ * @inproceedings{Chen:2016:XST:2939672.2939785,
+ *   author = {Chen, Tianqi and Guestrin, Carlos},
+ *   title = {{XGBoost}: A Scalable Tree Boosting System},
+ *   booktitle = {Proceedings of the 22nd ACM SIGKDD International Conference on
+ *                Knowledge Discovery and Data Mining},
+ *   series = {KDD '16},
+ *   year = {2016},
+ *   isbn = {978-1-4503-4232-2},
+ *   location = {San Francisco, California, USA},
+ *   pages = {785--794},
+ *   numpages = {10},
+ *   url = {http://doi.acm.org/10.1145/2939672.2939785},
+ *   doi = {10.1145/2939672.2939785},
+ *   acmid = {2939785},
+ *   publisher = {ACM},
+ *   address = {New York, NY, USA},
+ *   keywords = {large-scale machine learning},
+ * }
+ * @endcode
+ */
+template<typename LossFunction = SSELoss,
+         typename DimensionSelectionType =
+             mlpack::tree::MultipleRandomDimensionSelect,
+         template <typename> class NumericSplitType = XGBExactNumericSplit>
+class XGBoostTreeRegressor
+{
+ public:
+  //! Allow access to the underlying decision tree type.
+  typedef mlpack::tree::DecisionTreeRegressor<FitnessFunction,
+      NumericSplitType, mlpack::tree::AllCategoricalSplit,
+      DimensionSelectionType> DecisionTreeType;
+  /**
+   * Construct the xgboost without any training or specifying the number
+   * of trees.  Predict() will throw an exception until Train() is called.
+   */
+  XGBoostTreeRegressor();
+
+  /**
+   * Construct the xgboost forest and train.
+   */
+  template<typename MatType, typename ResponsesType>
+  XGBoostTreeRegressor(const MatType& dataset,
+                       const ResponsesType& responses,
```

Comment:
If you want to support categoricals, then I think all you would need to do is add another overload that takes a `data::DatasetInfo&` too.  I guess `SSELoss` would need some minor adaptation for that case, with a "regular" `Evaluate()` implementation.  (I think that should be pretty easy.)

Reply:
To be honest, I haven't given proper thought to it. I think we will also require a splitting strategy too that calculates and the sum of gradients and hessians accordingly (not 100% sure though). So, I would suggest we can leave it at this point. As I said in the previous comment, it won't be a big deal as I have considered it already.

One more thing, I haven't yet looked at how xgboost is doing it. They still have many open issues and PRs on it which I haven't had the chance to go through, but I would say we can at least take a look at them before just to be sure we are not overlooking something.

Let me know what you think. :)